lib/cartodb/backends/layer-stats/mapnik-layer-stats.js

```
@@ -11,17 +34,280 @@ MapnikLayerStats.prototype.is = function (type) {
+        tname += '[]';
+    }
+    return tname;
+}
+
+// columns are returned as an object { columnName1: { type1: ...}, ..}
+// for consistency with SQL API
+function formatResultFields(dbConnection, flds) {
+    flds = flds || [];
+    var nfields = {};
+    for (var i=0; i<flds.length; ++i) {
+      var f = flds[i];
+      var cname = dbConnection.typeName(f.dataTypeID);
+      var tname;
+      if ( ! cname ) {
+        tname = 'unknown(' + f.dataTypeID + ')';
+      } else {
+        tname = fieldType(cname);
+      }
+      nfields[f.name] = { type: tname };
+    }
+    return nfields;
+}
+
 MapnikLayerStats.prototype.getStats =
 function (layer, dbConnection, callback) {
-    var queryRowCountSql = queryUtils.getQueryRowCount(layer.options.sql);
-    // This query would gather stats for postgresql table if not exists
-    dbConnection.query(queryRowCountSql, function (err, res) {
-        if (err) {
-            return callback(null, {estimatedFeatureCount: -1});
-        } else {
-            // We decided that the relation is 1 row == 1 feature
-            return callback(null, {estimatedFeatureCount: res.rows[0].rows});
-        }
+    let aggrQuery = layer.options.sql;
+    let preQuery =  layer.options.sql_raw || aggrQuery;
+
+    let ctx = {
+        dbConnection,
+        preQuery,
+        aggrQuery,
+        metaOptions: layer.options.metadata || {}
+    };
+
+    // TODO: could save some queries if queryUtils.getAggregationMetadata() has been used and kept somewhere
+    // we would set queries.results.estimatedFeatureCount and queries.results.geometryType
+    // (if metaOptions.geometryType) from it.
+
+    // TODO: compute _sample with _featureCount when available
+
+    Promise.all([
+        _estimatedFeatureCount(ctx).then(
+            ({ estimatedFeatureCount }) => _sample(ctx, estimatedFeatureCount)
+                .then(s => mergeResults([s, { estimatedFeatureCount }]))
+        ),
+        _featureCount(ctx),
```

A question to understand it: If some of _featureCount, _aggrFeatureCount, _geometryType or _columns fails, we will return an error as a response. I know that the user must request for it expressly, but I am not sure if metadata should do to fail a request. What is your point of view?

---

It is not a reason to stop the PR and also, I am not saying to change it. Only it raises doubts.

---

IMHO, If the user requests for any specific metadata and an error happens, we should return that error because we weren't able to process the request completely.